### DISCUSSION TITLE
Modifier to mark interface implementation (`implement` complementing `override`)

### DISCUSSION THREAD

**Discussion by gatlex** [dotnet/csharplang#7764] [Language Ideas]
I suggest to introduce a modifier to mark a method as an _implementation_ of the `interface`.

DR: It greatly introduces maintainability. It is quite common in large projects and medium-sized refactorings, that a method is removed from an interface - but the (former) implementation remains as dead code.

Edit: Another solution for my goal would be to 'improve' on explicit interface implementation as detailed in [this comment](https://github.com/dotnet/csharplang/discussions/7764#discussioncomment-7907842).

----

Regarding naming I have no strong opinion. Two options that come to mind:
- Reuse `override`.  
  *Pro:* Keyword exists.  
  *Con:* Language lawyers will argue that this would not truly reflect the meaning as the method does not actually _override_ but _implements_ the interface method.
- New modifier `implement` or similar.

**Comment by HaloFour** [comment]
See also: #7210

For this specific case you could also use a custom attribute with an analyzer to warn/error on annotated methods that don't implicitly implement an interface member.  This is kind of how `@Override` works in Java, the compiler will complain if the member doesn't actually override anything.

**Reply by siegfriedpammer** [reply]
I strongly agree with you. This does not warrant the addition of a new keyword. Attributes + analyzers are fine. However I would love to see the proposal you linked to be implemented after we have DUs, roles/extensions and other powerful features, of course. It would provide more feature parity with other languages.

**Reply by gatlex** [reply]
Sorry, probably a newbie question - but how does this work in practice? Where does the attribute and the analyzer come from? (Since my context is a quite big company, it is not a viable option to manually extend the tool chain.)

Apart from that, attributes strike me as syntactically a bit combersome when in my view this should be the recommended style. Actually, I don't see why I am _obliged_ to specify `override` for (re-)implementing (abstract) class members while at the same time I am _disallowed_ to express the same intension for interface members.

Edit: Also, not sure what the connection to #7210 is.

**Reply by HaloFour** [reply]
Analyzers can be written in a separate project and you can take a dependency on it.  I don't think it would require additional configuration in your CI/CD pipeline.  They allow you to enforce custom static analysis rules as you see fit, including detecting when an annotated member does not correspond to a member on an implementing interface, or detecting dead code in general, etc.

C#/.NET offers two forms of implementation, implicit and explicit.  With implicit implementation, there is no hard dependency between the interface and the implementing type.  The type is considered to conform to the contract by virtue of compatible public members existing.  The runtime handles wiring up the implementation automatically.  This is unlike overriding, which must be explicit in the assembly metadata or the runtime considers it shadowing instead.

C# does have an explicit implementation syntax, but it's limited in that the member must be private and has a generated name that cannot be referenced from within the current type.  If you use that syntax, removing the member from the interface would definitely cause a compilation error.

The connection to the linked proposal is that I seek to expand on the explicit implementation syntax to allow more flexibility, such as changing the accessibility modifier or name of the member, while retaining that explicit connection to the interface member.

**Reply by gatlex** [reply]
Thank you very much for sharing your insights into the .NET runtime and explaining the connection to the proposal! (+1 now that I got it, btw.) To me it still seems, that the "runtime backend" found its way into the the "language frontend". Maybe the IL code is the same, but _conceptually_ I still don't see the difference in making the "matching" of members explicit (between derived and base class respectively between implementing class and implemented interface).

What explicit implementation syntax do you mean, I'm not aware of anything of the sort? (Maybe its a mitigation in some cases.)

A disadvantage of the analyzer approach then is, that it makes explicit changes to each project necessary (set up the analyzer and add the dependency). While it might be feasible sometimes (and I certainly will use it in selected projects), it's not exactly 'lightweight'.

**Reply by CyrusNajmabadi** [reply]
> What explicit implementation syntax do you mean, I'm not aware of anything of the sort? (Maybe its a mitigation in some cases.)

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/explicit-interface-implementation

**Reply by gatlex** [reply]
Ah, that one I actually know and now I think I know what you mean: The method in the example of the page
```c#
void IControl.Paint()
```
becomes in IL a private member with generated name, right?

That would be fine for me - but to use it, one has to upcast to the interface. Probably, that's what you mean by "[...] cannot be referenced from within the current type". I once stumbled over this, because I don't think it's what you would naivly expect (or only when ambiguities exist).

**Reply by HaloFour** [reply]
> Imo, implementing an interface clearly models an 'is a' relation

Or include a bridge method which you can call directly:

```cs
void IControl.Paint() => this.Paint();

public void Paint() { ... }
```

**Reply by gatlex** [reply]
> Or include a bridge method which you can call directly:

I know about this one. Imo its not practical, to laborious and much noise for nothing.

**Comment by the-avid-engineer** [comment]
You can use explicit interface implementations to get a similar effect

```cs
public interface IFoo
{
    void Bar();
}

public class Foo : IFoo
{
    void IFoo.Bar()
    {
        // implementation
    }
}
```

**Reply by KennethHoff** [reply]
The annoying part of it is that you can't cleanly call that method in other parts of the class

**Reply by the-avid-engineer** [reply]
Interfaces are meant to be a contract with the outside of the object, not with the inside of the object. Implicit interface implementations make it easy to ignore this fact, whereas explicit interfaces implementations do not

In that situation, I would go for something like this:

```cs
public interface IFoo
{
    void Bar();
}

public class Foo : IFoo
{
    private void SecretBar()
    {
        // do stuff
    }

    void IFoo.Bar()
    {
        SecretBar();
    }

    public void OtherBar()
    {
        SecretBar();

        // do other stuff
    }
}
```

I would have to agree with others about analyzer route if you want to avoid doing this kind of verbosity. A _language_ level feature that is only useful at the moment the code using it needs to be deleted seems a bit silly

**Reply by DavidArno** [reply]
> The annoying part of it is that you can't cleanly call that method in other parts of the class

Yes, for me this is what needs fixing with the language to effectively achieve what the OP wanted. We do not want to have to type the gnarly `void Baz => ((IFoo)this).Bar();` in order to access explicit interface implementations, so we don't use explicit interface implementations. As a result, the situation that the OP describes exists as we all use implicit interface implementations. Rather than fix the root cause, we then get "use attributes and write an analyzer" advice, which just piles more complication and problems onto the code.

Make the use of explicit interface implementations easier and this whole sorry house of cards issue would go away. Standard "pit of success" stuff, really.

**Reply by gatlex** [reply]
@the-avid-engineer: I agree with @DavidArno - imo your suggestion is just too much boilerplate to be relevant in practice. At least it's not feasible in my context (large existing code base with heterogenous dev environment) and, maybe even more importantly, it relys on the 'cooperation' and the discipline of the developer (which we have many).

> I would have to agree with others about analyzer route if you want to avoid doing this kind of verbosity.

Imo, analyzers are not feasible in many practical situations (mine being an example), same reasoning as above.

> A _language_ level feature that is only useful at the moment the code using it needs to be deleted seems a bit silly

To be honest, I don't see why. In my experience, code maintainability is a very valuable asset. And I don't see the difference to the `override` keyword. (I'm not sure about the IL, maybe something similar is necessary there; still, the compiler could detect that and add the keyword implicitly, I'm sure.)

**Reply by gatlex** [reply]
@DavidArno: Thanks for your eloquent support. 👍

Apart from your arguments, it's not feasible to add and maintain such an analyzer to all our projects (huge heterogenous code base).

But you shed light on another possiblity, namely to 'improve' the usability of explicit interface implementations:
```c#
// The following produces an compile error - why?! (Even worse, the message doesn't even hint to make the interface explicit.)
(new C()).Func(); // 'C' does not contain a definition for 'Func' and no accessible extension method 'Func' accepting a first argument of type 'C' could be found (are you missing a using directive or an assembly reference?)

// Works but is really cumbersome imo (even though it might be necessary in corner cases, which is acceptable, of course).
(new C() as I).Func();

interface I { void Func(); }
class C : I { void I.Func() {} }
```
This would solve my issue cleanly. I actually once stumbled (and fell) over this, but forgot about it again.

Imo, implementing an interface clearly models an 'is a' relation - which means that the interface methods be directly usable in the context of the class type. There may be corner cases like ambiguity in which the caller must decide which interface to use, but I think it's not a good idea to make the main use case so much more cumbersome. The current semantics should at most be an opt-in, I think.

**Reply by HaloFour** [reply]
You only need to develop an analyzer once, then you can share it across any codebase.  Being able to scan codebases to apply custom rules is one of the reasons that analyzers exist

Given that C# has functioned like this for 22 years it's very unlikely that this design philosophy around implicit implementation will be revisited.  Maybe explicit implementation enhancements will be considered, but otherwise your only option is likely to go the analyzer route.  That also lets you solve your problem immediately, whereas language features often take years even after they're championed.

**Reply by HaloFour** [reply]
> Imo, implementing an interface clearly models an 'is a' relation

Interface implementation is more accurately described as a "has-a" or "can-do" relationship.  They describe capabilities and a contract between the interface and the consumer of that interface.

**Reply by gatlex** [reply]
@HaloFour: Ok thanks again. I still deem an (optional) `implement` keyword worthy to be at least considered. It doesn't change the design philosophy, but as a dev, you can opt-in. But, as I said, I'd consider improving explicit implementation good enough (I edited the discussion body accordingly).

Regarding the newer post, I disagree with the "has a"; for me that would be a classical composite pattern. Maybe it's a "can do" - but that's just terminology for me; I'd still expect to be able to "do" it directly on the type. I suspect one can disagree on this point, though; still, a means to explicitly implement the interface and provide the same method in the class itself _is_ imo missing in the language.

Btw, the intention of this discussion is not a short-term solution for me, but to improve the language long-term.

**Reply by HaloFour** [reply]
Interfaces are intentionally a loosely relationship.  The contract is between the interface and the consumer, not the implementer.  That's why implementation can be satisfied through a number of alternative routes that do not include requiring members on a type.  For example, an interface member could be satisfied through a member defined on a base class.  The language doesn't even need to know or care about this, the runtime will wire up that relationship.  That is part of the philosophy of interfaces in C#, they're not intended to represent an "is-a" relationship like inheritance.  They are intended for composition scenarios.

Anyway, that's a philosophical discussion.  From a pragmatic point of view, the question is whether a language feature brings enough benefit to the language to be worth implementing.  I'd wager that just erroring/warning on a member that no longer implements an interface member is probably not enough, especially 22 years into the life of the language.

**Reply by DavidArno** [reply]
> Given that C# has functioned like this for 22 years it's very unlikely that this design philosophy around implicit implementation will be revisited. 

`InternalsVisibleTo` has also functioned the way it has for 22 years. And yet there's plenty of noise around that topic suddenly. So maybe after 22 years it's time to show explicit interface implementations some love too. 🤷

**Reply by HaloFour** [reply]
Close, `InternalsVisibleTo` was added in .NET 2.0, so only 18 years :)

Either way, they're different problems.  The purpose of exploring changes to the IVT space is to add more functionality, and I disagree with the direction that exploration seems to be taking for the moment.

I don't disagree that it would be beneficial to add more capabilities to explicit implementation, which is why I made my proposal to bring C#'s capabilities more inline with what the runtime offers (and VB.NET has supported since 1.0).  Notably, overrides offer that same functionality.

**Reply by CyrusNajmabadi** [reply]
> Imo, analyzers are not feasible in many practical situations (mine being an example)

Could you expand on that?  Why is it not feasible here?

**Reply by the-avid-engineer** [reply]
> > Imo, analyzers are not feasible in many practical situations (mine being an example)
> 
> Could you expand on that?  Why is it not feasible here?

I'm curious about this too. I have to think that if this _were_ to become a language feature, it would be behaving it exactly like an attribute + analyzer would. It's just a matter of being a native attribute/analyzer vs one you have to install manually.

I'm speculating, but the keyword would probably transform into a compiler-generated attribute, very similar to how `required` works, and then MSBuild would inspect the code for anything marked with that attribute and make sure that it implements at least one interface method, very similar to what an analyzer would do.

At a maximum, you have _two_ extra characters in the code (`[` and `]`)

```cs
public interface IFoo { void Bar(); }

// Attribute + Analyzer
public class FooWithAttr : IFoo {
    [Implement]
    public void Bar() { }
}

// Language
public class FooWithLangFeature : IFoo {
    public implement void Bar() { }
}
```

**Reply by gatlex** [reply]
What you say may be true for OSS projects, but not (necessarily) for commercial software development. I'd put it as more of a 'policitical' issue than a technical. As I said, we have a large code base spreaded over _many_ repos / projects / components. Devs are allocated across components as needed. For each repo (component), another person is responsible ("owner") and since it introduces dependencies, I as a dev have to convince the owner of each component. On the other hand, using new language features are no problem.

Apart from that it's also much more effort to deploy the analyzer to each project and add it as a dependency, which in turn I have to convince my product manager of. But that's a minor issue compared to the other.

My very personal opinion, apart from that, is that `[Implement]` is syntactically simply not pretty when compared to `public implementation void Bar()`.

**Reply by gatlex** [reply]
That said, if I were to choose between an `implement` keyword and 'improving' explicit implementation so that the 'bridge' @HaloFour mentioned in [this comment](https://github.com/dotnet/csharplang/discussions/7764#discussioncomment-7908320) is provided (with the same name) without noise (redundancy), I'd go for explicit implementation improvement. I don't see how attributes & analyzers help with that.

**Reply by HaloFour** [reply]
Analyzers were added to the Roslyn compiler specifically to enable developers (and companies) to design and enforce their own static analysis rules without requiring adding new language features.  Language design is a _massive_ bottleneck and it would be quite impossible for the language team to consider or implement a fraction of the proposals here.  If your company isn't willing to consider the options that are available, that's not going to change how the C# team approaches language design or encourage them to start taking scenarios easily solvable by the tools provided and baking them into the language.

**Reply by gatlex** [reply]
Maybe, but that's how life is. So we're stuck with less maintainable code for now, I guess.

**Reply by the-avid-engineer** [reply]
I don't think political blindness to the fact that upgrading the language version _also_ introduces new dependencies should dictate how C# evolves as a language.

IMHO the majority of the work on this, no matter which way it's done, would be updating to code to use it. The amount of time spent to update the language version or add one package is going to be vanishingly small in comparison

And if the keyword is `implementation`, the attribute would probably be `[Implementation]`.. not that it makes any difference on aesthetics of attribute vs keyword

**Reply by siegfriedpammer** [reply]
From personal experience with huge codebases using a mixture of programming languages and build tools, it is a lot easier to convince people to use/update a nuget package than convincing them to install a new version of visual studio or build tools on every build server and dev machine. Nuget packages are zero work after having them added to the project once on one machine.

This is even more the case for projects where C# is only used in a tiny fraction (compared to rest of the project) of the code.

**Reply by CyrusNajmabadi** [reply]
> I'd put it as more of a 'policitical' issue than a technical. 

That's definitely a non-starter for language design. Imagine if every time we designed a feature, we had to duplicate it all over again because some org says they couldn't use it for political reasons. 

Such reasons need to be hashed out by that group. The language is not the place to solve organizational dysfunction

**Reply by CyrusNajmabadi** [reply]
> My very personal opinion, apart from that, is that [Implement] is syntactically simply not pretty when compared to public implementation void Bar().

Note: I don't think this language feature could even work. 

Consider the following:

```
//Assembly1
interface I { void M(); }
class C { public void M() {} }

// Assembly2
class D : C, I { }
```

The compiler has no knowledge of D when compiling the first assembly. So it has no way to validate any sort of keyword. 

This would have to be an attribute that could be analyzed with full context of the entirety of downstream dependencies

**Reply by the-avid-engineer** [reply]
@CyrusNajmabadi If it were a language feature, it would probably be lowered to an attribute much like `required`

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi If it were a language feature, it would probably be lowered to an attribute much like required

:-)

My point here is that it can't really be a language feature. Specifically because the analysis and requirements it brings about can't actually be validated by the compiler on its own.  You have to know the entire set of downstream dependencies to actually validate things. Which is not something the language knows about or has ever created rules for.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

Right, an implementation of something like this, whether as a language feature or an analyzer, I think would have to work similarly to Java's `@Override` annotation.  It doesn't prevent implicit implementation, but if you mark a member and the compiler can't verify that the member implements an interface member, that would result in a diagnostic:

```cs
interface I {
    void M();
}

class A : I {
    public void M() { } // fine
}

class B : I {
    [Implements] public void M() { } // also fine

    [Implements] public void N() { } // diagnostic, N doesn't implement any interface members
}
```
